### DISCUSSION TITLE
Can you please add some examples on how to use Auto-Clone?

### DISCUSSION THREAD

**Discussion by oasaleh** [ZiggyCreatures/FusionCache#310] [Q&A] [answered]
Mainly while defining a named cache in program.cs and also when adding an entry to the cache.

Thanks!

**Comment by jodydonetti** [comment]
Hi @oasaleh , sorry for the delay but I just came back from (very late) summer vacations 😅

Yes, I surely can!

Have you seen the docs [here](https://github.com/ZiggyCreatures/FusionCache/blob/main/docs/AutoClone.md)? I'd like to know what kind of example you feel is missing.

Thanks!

**Comment by oasaleh** [comment]
There's no code on that page. For example, how to set it up with the serializer. Also, is there a default serializer I can use? If I use Microsoft's, will it get initialized every time I'm getting something from the cache, or only once during initialization of the cache?

**Reply by jodydonetti** [reply]
Got it, good points!
I assumed that since the serializer has always been part of the distributed cache part, people would know it: but you are right, somebody may not have used that part and a better intro is desirable.

Thanks, will do it soon!

**Reply by oasaleh** [reply]
Thank you so much. Yeah, I only use the in-memory cache, so I am not familiar with the distributed options/configs.

**Reply by jodydonetti** [reply]
Hi @oasaleh , I updated the [docs](https://github.com/ZiggyCreatures/FusionCache/blob/main/docs/AutoClone.md) based on your suggestions.

If you have a moment let me know what you think.

Thanks!

**Reply by oasaleh** [reply]
Thank you for adding the examples. I have a suggestion and a follow up question.

Suggestion:
I think a better terminology could be used to describe the code snippets. 
```csharp
// RETURNS A CLONE OF CACHED INSTANCE (ORIGINAL REMAINS UNCHANGED)
var person1 = cache.GetOrDefault<Person>("foo", options => options.SetAutoClone(true));
Console.WriteLine($"person1: {person1.Name}");
Console.WriteLine();

// RETURNS A CLONE OF CACHED INSTANCE (MODIFICATIONS AFFECT ONLY THE CLONE)
var person2 = cache.GetOrDefault<Person>("foo", options => options.SetAutoClone(true));
person2.Name = "Jane";
Console.WriteLine($"person1: {person1.Name}");
Console.WriteLine($"person2: {person2.Name}");
Console.WriteLine();

// RETURNS DIRECT REFERENCE TO CACHED INSTANCE (MODIFICATIONS AFFECT THE CACHE)
var person3 = cache.GetOrDefault<Person>("foo");
person3.Name = "Jim";
Console.WriteLine($"person1: {person1.Name}");
Console.WriteLine($"person2: {person2.Name}");
Console.WriteLine($"person3: {person3.Name}");
Console.WriteLine();

// RETURNS SAME REFERENCE TO CACHED INSTANCE
// MODIFICATIONS AFFECT BOTH person3 AND person4 AS THEY SHARE THE CACHED REFERENCE
var person4 = cache.GetOrDefault<Person>("foo");
person4.Name = "Joe";
```

I created a [PR ](https://github.com/ZiggyCreatures/FusionCache/pull/328) if you'd like to use that.

Question:
What about named caches? Do I add the serializer here in `Progam.cs` file?
```csharp
        builder
            .Services
            .AddFusionCache(Constants.MyCacheName)
            .WithDefaultEntryOptions(opt => { opt.Duration = TimeSpan.FromDays(10); });
```

Thank you.

**Reply by jodydonetti** [reply, marked as answer]
Hi @oasaleh 

> Question: What about named caches? Do I add the serializer here in `Progam.cs` file?
> 
> ```cs
>         builder
>             .Services
>             .AddFusionCache(Constants.MyCacheName)
>             .WithDefaultEntryOptions(opt => { opt.Duration = TimeSpan.FromDays(10); });
> ```

Yes, you set the serializer for each cache you setup, named or not:

```cs
builder
  .Services
  .AddFusionCache(Constants.MyCacheName)
  .WithSerializer(...) // HERE
  .WithDefaultEntryOptions(opt => { opt.Duration = TimeSpan.FromDays(10); });
```

Hope this helps!

**Reply by jodydonetti** [reply]
ps: if you've found this useful, please mark it as the answer, thanks!
